packages/babel-plugin-jest-hoist/src/index.js

```diff
@@ -8,11 +8,63 @@
 
 'use strict';
 
+function invariant(condition, message) {
+  if (!condition) {
+    throw new Error('babel-plugin-jest-hoist: ' + message);
+  }
+}
+
 const JEST_GLOBAL = {name: 'jest'};
 
+const idVisitor = {
+  ReferencedIdentifier(path) {
+    this.ids.add(path);
+  },
+};
+
 const FUNCTIONS = {
   mock: {
-    checkArgs: args => args.length === 1 && args[0].isStringLiteral(),
+    checkArgs: args => {
+      if (args.length === 1) {
+        return args[0].isStringLiteral();
+      } else if (args.length === 2) {
+        invariant(
+          args[1].isFunction(),
+          'The second argument of `jest.mock` must be a function.'
+        );
+
+        const ids = new Set();
+        args[1].traverse(idVisitor, {ids});
+
+        const outerScope = args[1].parentPath.scope;
+
+        for (let id of ids) {
+          const name = id.node.name;
+          let found = false;
+          let scope = id.scope;
+
+          while (scope !== outerScope) {
+            if (scope.bindings[name]) {
+              found = true;
+              break;
+            }
+
+            scope = scope.parent;
+          }
+
+          if (!found) {
+            invariant(
+              scope.hasGlobal(name) && (name === 'jest' || name === 'require'),
```

Comment:
What comes to mind right now: Object, Array, Map, WeakMap, Set, setInterval, Proxy, undefined, NaN, Infinity, RegExp, parseInt, parseFloat, and lots more. We will end up with a huge list that has to be maintained.

Plus I don't see how we improve user experience by checking all global variables. Let's review this example:

```js
import 'a';
import 'define-global-b';

jest.mock('a', () => b);
```

If we don't allow referencing global `b`, then the user will likely use `global.b` and might be confused why `b` is undefined. I would just add a note to the docs saying that mock factory will be executed when corresponding module is being imported.

What do you think?